### ISSUE TITLE
Upload/Download Speed counter stuck during the course of a 200+ file download

### ISSUE THREAD

**Issue by Searinox** [zlatinb/muwire#89]
This is happening on a different system than I am testing the current issues:

On this system I set some 200 files to download at once. At some point the download/upload counter at the bottom of the page froze. Downloads and uploads are still happening and in the respective tabs I can see them ongoing and completing but the counter is stuck. It's not just a case of speeds not returning to zero when there's no activity, it won't even increase even as downloads/uploads intensify.

Since you mentioned that these likely indicate hanged download/upload objects, I figured to report it. I will proceed to analyze this in visualvm like we did in #77 but you will have to tell me what to look for because there are now many ongoing downloads not just one.

**Comment by Searinox** [comment]
I paused all downloads and the counter didn't budge. Here's the stack trace:

1/2:
https://pastebin.com/raw/9W08LZCR

2/2:
https://pastebin.com/raw/kqGpCY2R

Was too big for a comment and even for a single pastebin.

**Comment by Searinox** [comment]
My bad. Here: https://pastebin.com/raw/x4ambSN7

So this is EventQueue:


"AWT-EventQueue-0" #20 prio=6 os_prio=0 cpu=503218.75ms elapsed=15271.22s tid=0x00000296f2795000 nid=0x77bc waiting on condition  [0x000000e2af1fe000]
   java.lang.Thread.State: WAITING (parking)
	at jdk.internal.misc.Unsafe.park(java.base@11.0.12/Native Method)
	- parking to wait for  <0x0000000404c231a8> (a java.util.concurrent.locks.AbstractQueuedSynchronizer$ConditionObject)
	at java.util.concurrent.locks.LockSupport.park(java.base@11.0.12/Unknown Source)
	at java.util.concurrent.locks.AbstractQueuedSynchronizer$ConditionObject.await(java.base@11.0.12/Unknown Source)
	at java.awt.EventQueue.getNextEvent(java.desktop@11.0.12/Unknown Source)
	at java.awt.EventDispatchThread.pumpOneEventForFilters(java.desktop@11.0.12/Unknown Source)
	at java.awt.EventDispatchThread.pumpEventsForFilter(java.desktop@11.0.12/Unknown Source)
	at java.awt.EventDispatchThread.pumpEventsForHierarchy(java.desktop@11.0.12/Unknown Source)
	at java.awt.EventDispatchThread.pumpEvents(java.desktop@11.0.12/Unknown Source)
	at java.awt.EventDispatchThread.pumpEvents(java.desktop@11.0.12/Unknown Source)
	at java.awt.EventDispatchThread.run(java.desktop@11.0.12/Unknown Source)

   Locked ownable synchronizers:
	- None

**Comment by zlatinb** [comment]
1.  If you can afford to lose those 200 downloads, cancel all of them, make sure they're cleared up from the UI, then take a heap dump with
```
jmap -dump:live,format=b,file=muwire-89.heap
```
If you can't afford to lose the downloads, take a heap dump anyway, it just means we'll need to do more work.

2. Open the heap with visualvm, look for the single instance of a `com.muwire.core.downloads.DownloadManager` class.
3. That class has a field called `downloaders`.  Focus on that object.  
4. That object will have a field called `baseCount` of type `long`.  If you cancelled all downloads it's value should be 0.  Otherwise it should be equal to the exact number of downloads.

If you cancelled the downloads the size of the downloaders Map isn't 0, or the size is not equal to the number of download tasks, we have a problem, and we'll investigate further.

**Comment by Searinox** [comment]
Aaaaaaaaaand it's gone. :|

Cancelling all download tasks unstuck the counter. It is now correctly showing 0 download speed and a dynamic upload speed based on an upload that is ongoing right now.

I do still have the heap dump saved from when it was stuck with 200+ downloads however.

**Comment by Searinox** [comment]
There is also the download-pumper:
![image](https://user-images.githubusercontent.com/5074487/137404470-1580d959-0ec6-4a8b-b94c-2237649f9924.png)

**Comment by Searinox** [comment]
As a side-question, are download list updates throttled when in view and disabled when not in the Download tab? I did notice some half-second UI freezes whenever multiple downloads would all choose to retry at the same time.

**Comment by zlatinb** [comment]
Just thought of something.  Some questions:
* Is automatic clearing of finished/cancelled downloads or uploads enabled on the node that experienced this bug?  
* Was the download or upload table sorted in any way?
* Were there any rows selected?

**Comment by Searinox** [comment]
Yes automatic clearing was enabled and the downloads list was sorted. No rows were selected. Trying this version now.

**Comment by Searinox** [comment]
I have to pause a substantial number of downloads for them to start going through now. Also yes the counter is stuck again, this time at all zero speeds.

**Comment by Searinox** [comment]
There exists an Upload Slots Per User setting in MuWire. Do clients communicate this value to one another? If they did, a downloading client may be able to organize its downloads to only accept a particular number of connections at a time. I think all these DownloadManagers need some kind of common load balancer object per each unique download source that can orchestrate this.

**Comment by Searinox** [comment]
Reproducing this is easy with the new folder downloads feature because it's not difficult to queue a handful of folders for download and the result can be a few hundred files easily. These in particular had only one source too so everything is being downloaded from there.

Unfortunately I do not have logging enabled but if it's like last time then I expect eventually the download instance that is holding this up will finish and will release the counter back into a working state. Because of this recovery tied to a download manager it seems more like a deadlock than an exception to me, but I'm not really one to opine since I don't know the code.

**Comment by Searinox** [comment]
This works. The counter is now moving under the reproduced conditions. Shall I close this and open another issue for congestion when downloading a large number of files from the same source? This is pretty straightforward to recreate and significantly hampers the basic mission of the program.

**Comment by zlatinb** [comment]
Definitely open a separate issue for the congestion.  As far as closing this one, I'm not 100% sure that we've found all possible causes for the counter getting stuck, but I guess this issue can be closed and then reopened if necessary.
